docs/adr/0006-process-for-onboarding-new-tenants.md

```diff
@@ -0,0 +1,299 @@
+
+```
+Root tenant
+ |
+ +-> Client Tenants
+ |     |
+ |     +-> Contoso
+ |     |
+ |     +-> Litware
+ |           +-> (Workflow storage configuration)
+ |           +-> (The Id of the WORKFLOW+Litware sub-tenant for the Workflow service to use)
+ |
+ +-> Service Tenants
+       |
+       +-> WORKFLOW
+       |     |
+       |     +-> WORKFLOW+Litware
+       |           +-> (Operations storage configuration)
+       |           +-> (The Id of the OPERATIONS+WORKFLOW+Litware sub-tenant for the Operations service to use)
+       |
+       +-> OPERATIONS
+       |     |
+       |     +-> OPERATIONS+WORKFLOW+Litware
+       |           +-> (FooBar storage configuration)
+       |
+       +-> FOOBAR
+```
+
+Next, the Workflow service continues its enrollment process by enrolling the new WORKFLOW+Litware tenant in its other dependency, FooBar. As with the Operations service enrolling OPERATIONS+WORKFLOW+Litware with FooBar, this does not result in any further tenants being created, just the FooBar config being attached to WORKFLOW+Litware:
+
+```
+Root tenant
+ |
+ +-> Client Tenants
+ |     |
+ |     +-> Contoso
+ |     |
+ |     +-> Litware
+ |           +-> (Workflow storage configuration)
+ |           +-> (The Id of the WORKFLOW+Litware sub-tenant for the Workflow service to use)
+ |
+ +-> Service Tenants
+       |
+       +-> WORKFLOW
+       |     |
+       |     +-> WORKFLOW+Litware
+       |           +-> (Operations storage configuration)
+       |           +-> (The Id of the OPERATIONS+WORKFLOW+Litware sub-tenant for the Operations service to use)
+       |           +-> (FooBar storage configuration)
+       |
+       +-> OPERATIONS
+       |     |
+       |     +-> OPERATIONS+WORKFLOW+Litware
+       |           +-> (FooBar storage configuration)
+       |
+       +-> FOOBAR
+```
+
+This completes Litware's enrollment for the Workflow service. As can be seen, this has resulted in multiple service-specific Litware tenants being created but Litware is never explicitly made aware of the existence of these tenants, nor is it able to use them directly. They are used by their parent services to make calls to their dependencies _on behalf_ of the Litware tenant.
+
+However, there is a further step: Litware also needs to be enrolled in the Operations service. At present, it is able to indirectly use the service via Workflow, but also needs to be able to use it directly. So, the process we went through for Workflow is repeated: the management API requests the description of the configuration that must be provided to enroll Litware in the Operations service - which is storage configuration for Operations and FooBar - and then supplies that configuration to the Operations service's enrollment endpoint.
```

Comment:
However...can Litware be completely unaware of Workflow's use of Operations?

Practically speaking, the fact that Workflow hands out what are quite clearly URLs into the Operations service means that even if we don't tell Litware what's happening, they will most likely deduce that Operations is involved here. But there's a deeper issue here.

Hypothetically, yes Litware could be completely unaware of it in the same way that they could also be completely unaware that Workflow uses CosmosDB. That's also an implementation decision that Workflow could choose to dispense of in the future, much as it could dispense of its use of Operations. But if we are exposing to customers the fact that Workflow uses CosmosDB (and if we're contemplating a 'bring your own storage' model, then we definitely are) I don't see that it makes sense to do that and yet not expose the fact that this also means they're using Azure Blob Storage via the Storage service.

Reply:
I was expecting that we would expose that they are using blob storage in this scenario. In the workflow -> operations example, when enrolling into workflow I was expecting to have to gather config for both services - see line ~99:

_Firstly, we will request the list of required configuration to enroll a tenant for use with Workflow. Workflow "knows" it needs storage configuration, and because it's dependent on Operations and FooBar, it also requests the required configuration for those services. Operations does the same with FooBar, so the total list of required configuration for Workflow is the sum of those four things: Workflow storage config, Operations storage config, FooBar storage config when invoked from Workflow, and FooBar storage config when invoked from Operations._

If they are using Marain tenanted storage, they won't actually care and we will just use defaults for all of those things. If they are bringing their own, we will have to go through this with them as part of the onboarding/enrollment process, which will necessarily be more interactive due to the need to set up permissions, get connection strings, etc, etc.

So, whilst they _could_ use Workflow in blissful ignorance of the fact that it's using Operations , we aren't trying to hide what's going on for them and in some cases we might actually require them to know.

Reply:
I'm not saying we are "hiding" it from them exactly, though the fact that we are giving them the opportunity to supply us with a Blob storage account that we want to use for storing long operation details doesn't necessarily indicate that we exposing the fact that we are using "Marain Operations" to do that. In fact, I probably *would* avoid making that clear.

Also, per IanG

> Practically speaking, the fact that Workflow hands out what are quite clearly URLs into the Operations service means that even if we don't tell Litware what's happening, they will most likely deduce that Operations is involved here.

There is no reason we should hand out URLs that are "quite clearly into the operations service"; providing an APIM facade over that would probably be better (so that we don't have the "on-behalf" tenant ID in an external URL, for example). That way we don't leak that implementation detail, and are protected against versioning issues.

Reply:
I think the wording now reflects that!

Reply:
>There is no reason we should hand out URLs that are "quite clearly into the operations service"; providing an APIM facade over that would probably be better (so that we don't have the "on-behalf" tenant ID in an external URL, for example). That way we don't leak that implementation detail, and are protected against versioning issues.

I can think of one reason: hiding the tenant via the APIM layer adds a whole load of extra complexity in APIM. Imagine a world where I'm using 3 Marain services, each of which uses Operations as an implementation detail and hands out URLs to the client to operations in progress. There will now be 3 distinct on behalf of tenant ids. Somehow, the logic in APIM that conceals these needs to work out for each incoming request to get the status of one of these operations what the tenant id is. I don't know how it's going to do that if we stripped it out. I guess we could encrypt the URL path when rewriting the Location header of any response and decrypt it when the relevant GET requests come back in, but this seems like quite a lot of extra complexity to bake into the APIM layer to me.

More broadly, solutions of the form "We'll just handle that in APIM" may look architecturally satisfying, but I think that's only because they remove the problems from view. My concern is that the APIM layer will become a festering rat's nest.